[KSTORAGE-830] Expand Fault Coverage of Kibosh

yyu1993 opened the pull request:
- Added unwritable and data corrupt faults to Kibosh.
-- Write_delay fault will delay a write call
-- Unwritable fault will intercept write system calls and return a user specified error code when injected.
-- Read_corrupt fault will read and return corrupted data when injected.
-- Write_corrupt fault will write corrupt data to file when injected.
- Data corrupt modes are standardized as constants for both read and write corrupt faults.
- Implemented fraction feature for read_delay fault
- Added suffix parameter for all faults
- Some minor bug fix.
- Testing can be done by running Kibosh in a linux docker container. A test framework will be added in the next PR.

yyu1993 committed:
Add write_corrupt fault and standarize corrupt modes for read and write corrupt faults.

yyu1993 committed:
Merge branch 'KSTORAGE-830' of github.com:yyu1993/kibosh

yyu1993 committed:
add random seed option

yyu1993 committed:
add file type parameter to all faults

yyu1993 committed:
add examples to README

yyu1993 committed:
add remove fault to readme

yyu1993 committed:
add process to clear page cache every 5 seconds

RamanVerma reviewed (commented):
@yyu1993  I have left a few comments and am going over the changes.

RamanVerma commented:
Please update the test strategy as well. Leave a note on the PR how we intend to test these changes.

RamanVerma commented on file.c:
```diff
@@ -276,12 +348,103 @@ int kibosh_release(const char *path UNUSED, struct fuse_file_info *info)
+                    buf_write[i] = buf[i];
+                }
+            }
+            buf_write[ret] = '\0';
```
We are using `buf_write` only when we want to debug. Is it possible to check the DEBUG macro is enabled before we do this copy operation. Also, this debug message makes more sense when we are writing zeroes as part of corruption. When we write a random integer, it is not useful.
User may supply a flag when they need to see the buffer data. Most of the times, it should be ok to tell that we corrupted the buffer.

Similar comment for the `buf_read` buffer at read function

yyu1993 replied:
A flag was added to corruption faults to let user decide whether to keep the corrupted data for reference.

RamanVerma commented on main.c:
```diff
@@ -212,6 +213,20 @@ int main(int argc, char *argv[])
+    }
+
+    /* Start a process to clear page cache every 5 seconds. */
+    int sret = system("while true; do sleep 5; sudo sh -c \"echo 1 > /proc/sys/vm/drop_caches\"; done &");
```
Could you please explain why do we need to do this ? A comment will be good.

yyu1993 replied:
In order to make Kibosh compatible with Kafka, we had to remove the direct_io option, which allows Kibosh to retain cache.
This leads to two consequences:
1. page cache might build up
2. for read_delay fault, only the first read call is affected by the fault, the subsequent read calls can access the cache instead.
So the process is intended to clear the page cache periodically to mitigate those two problems.

dhruvilshah3 replied:
@yyu1993 could you please add this comment in the code?

yyu1993 replied:
Sure

RamanVerma reviewed (commented):
Few more comments.

RamanVerma commented:
Please make sure that if you had to figure out operational things like deployment and configs, you should add them to the project's ReadMe file as part of this PR

yyu1993 committed:
Add examples and deployment instructions in README.md.
Add flag for storing data in corruption faults.
Style change as suggested by comments.

yyu1993 committed:
Merge pull request #1 from yyu1993/prcomment

Address comments

yyu1993 commented on fault.c:
```diff
@@ -49,6 +97,11 @@ static struct kibosh_fault_unreadable *kibosh_fault_unreadable_parse(json_value
     }
+    file_type_obj = get_child(obj, "file_type");
+    if ((!file_type_obj) || (file_type_obj->type != json_string)) {
+        INFO("kibosh_fault_unreadable_parse: No valid \"file_type\" field found in fault object.\n");
```
Function macro was added to replace the function names

yyu1993 commented on fault.c:
```diff
@@ -157,24 +233,289 @@ static int kibosh_fault_read_delay_check(struct kibosh_fault_read_delay *fault,
         return 0;
     }
-    milli_sleep(fault->delay_ms);
+    if (strlen(fault->file_type) > strlen(path) ||
```
file_type is intended to specify the extension of the file. In Kibosh, I match the file path to the file_type from right to left, so this can be interpreted as a file path suffix as well.

yyu1993 commented on fault.c:
```diff
@@ -157,24 +233,289 @@ static int kibosh_fault_read_delay_check(struct kibosh_fault_read_delay *fault,
+                    "\"file_type\":\"%s\", "
+                    "\"mode\":%d, "
+                    "\"fraction\":%g}",
+                    KIBOSH_FAULT_TYPE_READ_DELAY,
```
fixed, thanks for pointing that out.

yyu1993 commented on fault.c:
```diff
@@ -157,24 +233,289 @@ static int kibosh_fault_read_delay_check(struct kibosh_fault_read_delay *fault,
+                    "\"file_type\":\"%s\", "
+                    "\"mode\":%d, "
+                    "\"fraction\":%g}",
+                    KIBOSH_FAULT_TYPE_READ_DELAY,
```
fixed.

yyu1993 commented on fault.c:
```diff
@@ -391,12 +756,17 @@ int faults_check(struct kibosh_faults *faults, const char *path, const char *op)
-        if (ret)
-            return ret;
+        int err_code = kibosh_fault_base_check(*iter, path, op);
+        if (err_code)
```
fixed.

yyu1993 commented on file.c:
```diff
@@ -236,6 +237,77 @@ int kibosh_read(const char *path UNUSED, char *buf, size_t size, off_t offset,
+                        file_type = corrupt_fault->file_type;
+                    }
+                }
+
```
For the corruption faults, fraction can be zero. In such case, pos will equal to ret and nothing will be done to the buffer. For the file_type, once the fault is successfully parsed into Kibosh, file_type can never be NULL, it can be "" though.

yyu1993 committed:
add unmount instructions

dhruvilshah3 commented on fault.h:
```diff
@@ -24,6 +24,20 @@
+/**
+ * Generate a random double between 0 and 1.0
+ */
+#define RAND_FRAC ((double)rand()/(double)RAND_MAX)
```
Can we turn this into a function instead?

yyu1993 replied:
Added a random_fraction function in util to replace RAND_FRAC

dhruvilshah3 commented on fault.c:
```diff
@@ -391,12 +770,19 @@ int faults_check(struct kibosh_faults *faults, const char *path, const char *op)
+        int err_code = kibosh_fault_base_check(*iter, path, op);
+        if (err_code) {
+            // get the smallest err_codes
+            // i.e. when unreadable and read_corrupt faults both present, we want to execute unreadable
```
This comment does not seem very accurate. Could you elaborate a bit on what this piece of code is doing?

yyu1993 replied:
The reason I added this is to handle the case where two read faults are injected for the same path and file_type. The error code for unreadable faults are from 1 to 122, and error code (mode) for corrupt faults are greater than 1000. So if both faults are presented, unreadable will be applied. Without the check, Kibosh would arbitrarily inject the first fault it finds, so the result is not deterministic.

dhruvilshah3 commented on fault.h:
```diff
@@ -24,6 +24,20 @@
+#define CORRUPT_RAND 1001
+#define CORRUPT_ZERO_SEQ 1100
+#define CORRUPT_RAND_SEQ 1101
+#define CORRUPT_DROP 1200       // this will silently drop data from write calls
```
What is the reason for picking numbers like these? Could this just be an enum?

yyu1993 replied:
Referring to the previous comment, I picked large numbers to distinguish corruption faults from unreadable/unwritable faults. An enum would work as well.

dhruvilshah3 commented on fault.h:
```diff
@@ -89,6 +128,127 @@ struct kibosh_fault_read_delay {
+     * 0 means do not store data.
+     * Greater than 0 means store data to log.
+     */
+    int store_data;
```
What does it mean to have this set for a read operation?

yyu1993 replied:
This flag is added to prevent unnecessary logs. In corruption faults, the corrupted data would be written to the logs for future references. User can set this flag to 0 to not store the data (can be large). Setting the flag to greater than 0 means writing corrupted data to Kibosh logs.

dhruvilshah3 commented on fault.h:
```diff
@@ -89,6 +128,127 @@ struct kibosh_fault_read_delay {
+     * Mode of byte corruption.
+     * 1000 -> zeros (default)
+     * 1001 -> random values
+     * 1100 -> sequential zeros
```
Could we elaborate a bit on the difference between the `sequential` and non-sequential variants?

yyu1993 replied:
Sequential corruption will only corrupt the continuous portion of the file near the end, depending on the fraction set on the fault. Non-sequential variants will corrupt random positions in a file. So sequential corruption was intended to simulate a power fail scenario where the end of a file might not be written to disk correctly. Non-sequential corruption corrupts random bytes of the file and can mimic other disk corruptions.

dhruvilshah3 replied:
I understand that. I meant that we should add a comment in the code about the difference.

dhruvilshah3 replied:
@yyu1993 could you add this comment to the code?

yyu1993 replied:
Comment was added.

dhruvilshah3 commented on file.c:
```diff
@@ -276,12 +349,104 @@ int kibosh_release(const char *path UNUSED, struct fuse_file_info *info)
 }
 
+_Pragma("GCC diagnostic push")
+_Pragma("GCC diagnostic ignored \"-Wdiscarded-qualifiers\"")
```
Is this needed?

yyu1993 replied:
When corrupting write calls, the write buf needed to be changed in order to write corrupted data to disk. However, the buf pointer is a const, so changing it will raise a discarded-qualifier warning from the compiler. The pragma is only to suppress the warning during compilation.

cmccabe replied:
I'd really rather not modify a const buffer.  It's technically undefined behavior.  Let's just copy the buffer in the case where we need to do this.  You should be able to copy it on the stack since it's a few KB at most

dhruvilshah3 reviewed (commented):
Thanks for the PR, @yyu1993. I left a few initial comments.

yyu1993 commented on fault.c:
```diff
@@ -38,46 +85,56 @@ static struct kibosh_fault_unreadable *kibosh_fault_unreadable_parse(json_value
+        goto error;
+    }
+    file_type_obj = get_child(obj, "file_type");
+    if ((!file_type_obj) || (file_type_obj->type != json_string)) {
```
Originally, I made file_type optional, however, since Kibosh uses Json strings to specify faults, it seems more robust to make each Kibosh Json string having fixed fields. If user set file_type to "", it still means that Kibosh will apply fault to all the files. In the python wrapper for Ducktape, the type_file field is optional with default values, so it is still easy to use for the users.

yyu1993 commented on conf.c:
```diff
@@ -36,6 +36,8 @@ struct fuse_opt kibosh_command_line_options[] = {
      KIBOSH_CONF_OPT("--target %s", target_path, 0),
      KIBOSH_CONF_OPT("--control-mode=%o", control_mode, 0600),
      KIBOSH_CONF_OPT("--control-mode %o", control_mode, 0600),
+     KIBOSH_CONF_OPT("--random-seed=%d", random_seed, 0),
```
I feel that random-seed is more appropriate as it's the seed for the rand() function?

yyu1993 committed:
change RAND_FRAC to a function in util

yyu1993 committed:
fix ;

yyu1993 committed:
change corrupt constants to enum

yyu1993 committed:
improve logging

yyu1993 committed:
Merge pull request #2 from yyu1993/logging

improve logging

yyu1993 committed:
add std=c99 in cmakelists

yyu1993 committed:
Merge branch 'master' of github.com:yyu1993/kibosh

yyu1993 committed:
add std=c99 to cmakelists

yyu1993 committed:
fix for loop for c99

yyu1993 committed:
fix unreadable return code

dhruvilshah3 commented on file.c:
```diff
@@ -232,6 +231,15 @@ int kibosh_read(const char *path UNUSED, char *buf, size_t size, off_t offset,
     ret = pread(file->fd, buf, size, offset);
+
+    // clear page cache
+    int sret = system("sudo sh -c \"echo 1 > /proc/sys/vm/drop_caches\"");
```
Why do we do this here? A comment would be useful.

yyu1993 replied:
I experimented with cleaning the cache from read calls, but the result was not very predictable. The cache was not cleared after read call returned unless we spawn a background process here. So I switched back to having one background process spawned by main to clear the cache. I changed the interval to every half second.

dhruvilshah3 commented on fault.h:
```diff
@@ -27,16 +27,13 @@
+    CORRUPT_ZERO_SEQ = 1100,    // replace sequential bytes at the end of the file with \0
+    CORRUPT_RAND_SEQ = 1101,    // replace sequential bytes at the end of the file with random char value
+    CORRUPT_DROP = 1200,        // silently drop bytes at the end of the file
+} corrupt_mode;
```
Can you add a comment for why we pick these specific numbers here? Also, is there a reason these codes are not sequential?

yyu1993 replied:
Sure! I left some gap between the numbers to group sequential corruptions and random corruptions. We may add more corruption modes later for sequential bytes or random bytes.

yyu1993 committed:
change clear cache interval to .5 sec

yyu1993 committed:
add comments for corruption modes

yyu1993 committed:
add write_delay fault

yyu1993 committed:
change clear cache interval to 1 sec

dhruvilshah3 commented on fault.c:
```diff
@@ -30,6 +29,54 @@
+/**
+ * fault free functions
+ */
+static void kibosh_fault_unreadable_free(struct kibosh_fault_unreadable *fault)
```
@yyu1993 could you address this comment?

yyu1993 replied:
The repeated fault free functions were intended to allow faults to have different parameters. Right now all faults have prefix and file_type that need to be freed, but we can also add some parameter to an individual fault and customize its free function. We already have a kibosh_fault_base_free function that takes in a generic fault object and maps it to its specified free function.

dhruvilshah3 commented on file.c:
```diff
@@ -236,8 +236,80 @@ int kibosh_read(const char *path UNUSED, char *buf, size_t size, off_t offset,
+
+                    case CORRUPT_RAND_SEQ:
+                        for (; pos < ret; ++pos) {
+                            memset(buf+pos, (int) round(random_fraction() * 255.0), 1);
```
nit: could we do this in a single `memcpy` call instead? Perhaps something like this:

```
int numBytes = ret - pos;
char *randomBuf = malloc(numBytes);
int i;
for (i = 0; i < numBytes; i++)
  randomBuf[i] = round(random_fraction() * 255.0), 1);
memcpy(buf+pos, randomBuf, numBytes);
```

This could be similarly be done for other cases.

yyu1993 replied:
edited as suggested.

dhruvilshah3 reviewed (approved):
Thanks for the PR @yyu1993. LGTM. There are a few minor comments, would be good to address them before you merge.

yyu1993 committed:
add comments and change memset to memcpy

yyu1993 committed:
fix

cmccabe commented on README.md:
```diff
@@ -6,12 +6,41 @@ Kibosh acts as a pass-through layer on top of existing filesystems.  Usually,
+C compiler.  We also depend on the development libraries for FUSE.  Pkg-config is
+used by configure script to gather paths of required libraries.
+
+Dependencies can be installed with the following commands.  Run configure before
```
Probably better to shorten this as "in Ubuntu, run `apt-get cmake libfuse-dev fuse pkg-config` to install the dependencies

yyu1993 replied:
Edited as suggested.

cmccabe commented on README.md:
```diff
@@ -28,10 +57,14 @@ To run the tests for Kibosh, type:
 
-To run Kibosh, you supply the mount point as the first argument, and the
-directory which you would like to mirror as the second argument.
+To run Kibosh, you supply a mirror as the mount point as the first argument, 
```
how about "you supply a mirror directory..."

yyu1993 replied:
Edited as suggested.

cmccabe commented on conf.c:
```diff
@@ -36,6 +36,8 @@ struct fuse_opt kibosh_command_line_options[] = {
      KIBOSH_CONF_OPT("--target %s", target_path, 0),
      KIBOSH_CONF_OPT("--control-mode=%o", control_mode, 0600),
      KIBOSH_CONF_OPT("--control-mode %o", control_mode, 0600),
+     KIBOSH_CONF_OPT("--random-seed=%d", random_seed, 0),
```
should it be initialized to `time()` so that it's unpredictable by default?

yyu1993 replied:
Originally the default seed was set to time(0) in main.c. I think it makes more sense to set it here. Edited as suggested.

cmccabe commented on fault.h:
```diff
@@ -89,6 +133,143 @@ struct kibosh_fault_read_delay {
+     * 1101 -> sequential random values
+     * 1200 -> silently drop a fraction of bytes at the end of buffer
+     */
+    int mode;
```
this should use the enum type defined earlier, right?

yyu1993 replied:
edited as suggested.

cmccabe commented on fault.h:
```diff
@@ -89,6 +133,143 @@ struct kibosh_fault_read_delay {
+     * 1101 -> sequential random values
+     * 1200 -> silently drop a fraction of bytes at the end of buffer
+     */
+    int mode;
```
should use enum type defined earlier right

yyu1993 replied:
Edited as suggested.

cmccabe commented on fault.h:
```diff
@@ -89,6 +133,143 @@ struct kibosh_fault_read_delay {
+    int count;
+
+    /**
+     * The fraction of bytes to be corrupted.
```
should be documented as between 0.0 and 1.0 inclusive right

yyu1993 replied:
Added comment.

cmccabe commented on file.c:
```diff
@@ -236,8 +236,84 @@ int kibosh_read(const char *path UNUSED, char *buf, size_t size, off_t offset,
+                        break;
+
+                    // CORRUPT_ZERO
+                    default:
```
if this is just intended for CORRUPT_ZERO then specify that, and don't have a default case.  the compiler will warn if a case is not covered for an enum type (and you should be using the enum type, as described earlier)

yyu1993 replied:
Removed the default option.

yyu1993 committed:
change README as suggested

cmccabe commented on main.c:
```diff
@@ -212,6 +215,26 @@ int main(int argc, char *argv[])
+     * 1. preventing page cache from building up
+     * 2. force read and write calls to access the filesystem instead of the page cache to ensure fault can be injected
+     */
+    int sret = system("while true; do sleep 1; sudo sh -c \"echo 1 > /proc/sys/vm/drop_caches\"; done &");
```
If this is really necessary then let's create a thread to do it, rather than using bash.

yyu1993 replied:
Changed to child process as suggested.

yyu1993 committed:
change clear cache process to child process

yyu1993 committed:
Merge pull request #3 from yyu1993/prcomment812

change clear cache process to child process

cmccabe commented:
LGTM
